### PR TITLE
feat: resent failed emails

### PR DESCRIPTION
**Pull request by ReubenRobertsHEE**
Dependent on https://github.com/Health-Education-England/TIS-OPS/pull/775

TIS21-5746
TIS21-5747
TIS21-5745

### CODE REVIEW

File: src/main/java/uk/nhs/tis/trainee/notifications/service/EmailService.java [hunk 1]
```diff
@@ -120,8 +121,89 @@ public void sendMessageToExistingUser(String traineeId, NotificationType notific
...
+      historyService.save(history);
+
+      log.info("Sent template {} to {}.", templateName, recipient);
+    } else {
+      try {
+        log.info("For now, just logging mail to '{}' with subject '{}' and content '{}'", recipient,
+            helper.getMimeMessage().getSubject(), helper.getMimeMessage().getContent().toString());
+      } catch (IOException e) {
+        //should not happen
+        log.info("For now, just logging mail to '{}' with subject '{}' and unparseable content",
```
Comment (ReubenRobertsHEE) [inline src/main/java/uk/nhs/tis/trainee/notifications/service/EmailService.java:150]: I don't know how to test this unfortunately 😖

Comment (Judge40) [reply src/main/java/uk/nhs/tis/trainee/notifications/service/EmailService.java:150]: You can verify that the IOException is swallowed and doesn't trigger any other type of exception

```java
@Test
void coverage() throws IOException, MessagingException {
  MimeMessage mimeMessage = mock(MimeMessage.class);
  when(mailSender.createMimeMessage()).thenReturn(mimeMessage);

  when(mimeMessage.getContent()).thenThrow(IOException.class);

  assertDoesNotThrow(() -> service.sendMessage(TRAINEE_ID, RECIPIENT, NOTIFICATION_TYPE, "v1.2.3",
      new HashMap<>(), null, true));
}
```

File: src/test/java/uk/nhs/tis/trainee/notifications/service/HistoryServiceIntegrationTest.java [hunk 2]
```diff
@@ -169,15 +169,15 @@ void shouldSortFoundNotificationsBySentAtWhenMultipleFound() {
...
     Instant now = Instant.now().truncatedTo(ChronoUnit.MILLIS);
     service.save(new History(null, tisReferenceInfo, FORM_UPDATED, recipientInfo, templateInfo,
-        now, now, SENT, null));
+        now, now, SENT, null, null));
 
     Instant before = SENT_AT.minus(Duration.ofDays(1));
     Instant after = SENT_AT.plus(Duration.ofDays(1));
     service.save(new History(null, tisReferenceInfo, FORM_UPDATED, recipientInfo, templateInfo,
-        before, after, SENT, null));
+        before, after, SENT, null, null));
```
Comment (Judge40) [inline src/test/java/uk/nhs/tis/trainee/notifications/service/HistoryServiceIntegrationTest.java:177]: Having to chase these changes all over the app makes me regret using records at times 🤦🏻 

I think Lombok can add builders to records which might give us the best of both worlds (something for the backlog!).

**Review by Judge40 — APPROVED** [review]
Nothing that I thought was worth blocking the approval but there were a couple of minor changes you could make.
Happy enough with you leaving as-is or fixing in a follow-up

File: src/main/java/uk/nhs/tis/trainee/notifications/service/EmailService.java (hunk 1 above)
Comment (ReubenRobertsHEE) [reply src/main/java/uk/nhs/tis/trainee/notifications/service/EmailService.java:150]: Ah, mocking the MimeMessage, that I didn't think of! 🌮

File: src/test/java/uk/nhs/tis/trainee/notifications/service/HistoryServiceIntegrationTest.java (hunk 2 above)
Comment (ReubenRobertsHEE) [reply src/test/java/uk/nhs/tis/trainee/notifications/service/HistoryServiceIntegrationTest.java:177]: ✔️ . It looks like IntelliJ didn't support that properly initially, but it was fixed a year ago or so.

**Review by Judge40 — APPROVED** [review]

**Pull request merged**